Jenkinsfile

```diff
@@ -0,0 +1,307 @@
+    stage('CheckPrecheck') {
+      agent { label "master" }
+      when {
+        expression { DID_SKIPTEST == 'false' }
+      }
+      steps {
+        echo "Warning, skipped CI skip check"
+      }
+    }
+
+    // We will always need some of these to run the static checks.
+    // We need to git clone the:
+    // - tests repo so we can use some of its CI scripts
+    // - rutime repo to get the versions.yaml file
+    // - ci repo to get the Jenkins test matrix YAML
+    stage('Setup test repo environment') {
+      agent { label "master" }
+      steps {
+        sh '''
+          git clone "https://${tests_repo}.git" "${WORKSPACE}/${tests_repo_dir}"
+          git clone "https://${runtime_repo}.git" "${WORKSPACE}/${runtime_repo_dir}"
+          git clone "https://${ci_repo}.git" "${WORKSPACE}/${ci_repo_dir}"
+        '''
+      }
+    }
+
+    // Always run the static checks, even if we have a 'skip-ci' label.
+    // Later we could add a 'really-skip-ci' label if we needed etc. to
+    // not even run the static checks??
+    // If the PR does not pass the static checks,
+    // then the rest of the CI checks will be skipped by Jenkins.
+    stage('Static checks') {
+      agent { label "master" }
+      steps {
+        script {
+          echo "In static check"
+        }
+      }
+    }
+
+//    stage('examine the rootdir') {
+//      agent { label "master" }
+//      steps {
+//        script {
+//          sh "pwd"
+//          sh "ls"
+//        }
+//      }
+//    }
+
+    // Load up our test matrix data from the YAML file so we can tell
+    // which tests are to be run on which distro/jobs
+    stage('Load test matrix data') {
+      agent { label "master" }
+      steps {
+// FIXME - if we are test the CI repo, then we need to pick the config file up from the
+// Jenkins checked out PR (root of WORKSPACE), and not the golang checked out
+// ci repo. Code that up when we have access to the ghprbGhRepository var...
+// FIXME
+// // By default we run the tests using the git checkout out test repo
+// testcode_rootdir = env.ci_repo_dir
+// if ( env.ghprbGhRepository != null ) {
+//   // but, if this PR is from the tests repo, use the Jenkins checked out
+//   // source/branch of that instead, in case the PR is updating the test code.
+//   if ( env.ghprbGhRepository == env.ci_repo_name ) {
+//     testcode_rootdir = ""
+//   }
+// }
+        //dir("${testcode_rootdir}") {
+          script {
+            sh "pwd"
+            sh "ls"
+            testmatrix = readYaml file: "${testmatrix_file}"
+            echo "testmatrix looks like ${testmatrix}"
+         // }
+        }
+      }
+    }
+
+    // Run the CI on the primary distro (note, that does not mean this is our
+    // favoured distro, it just means we needed to pick one as the primary
+    // smoke test). If the primary distro does not pass, the rest of the distro
+    // tests will be skipped by Jenkins.
+    stage('Ubuntu18.04') {
```

Comment:
Nit: I find this hard to read - could you add a dash or something to make it `Ubuntu-18.04` maybe?

Reply:
Whatever - the whole thing is just an example right now. The names probably want to match the node/agent names on the master node, so we can then define which nodes run which bits from the Jenkinsfile. In fact, in theory we could code all of this into the YAML file I think, and have the Jenkinsfile a lot more 'dynamic' (but, that also scares me a touch ;-) )

Reply:
tbh, I'd much rather we encoded this logic in YAML as that a *lot* easier to read than all these horrid heavily nested braces imho ;)

Reply:
I think there are two things we could do:

1) If we go with the 'one pipeline job for each repo', then the Jenkinsfile could iterate the map of distros it finds from the YAML, and launch them. We could even mark which ones were 'primary' gating ones to run first.
1) If we go with 'one pipeline job per distro', then each of those pipelines can define what distro they are for in their env, and then the Jenkinsfile can use that var to pick out the correct section of the YAML file to execute. That should be easy.

and then,
1) We could actually code up the test commands to run, or the make commands to run, in the YAML. So, rather than `docker: true`, we could have things like `docker: "make docker"` - and the Jenkinsfile could literally run what we have in the YAML file - which could give us a lot of flexibility in the future.